### PR TITLE
udp: add router for UDP proxy

### LINKED ISSUE #17652: Route UDP packets to different upstreams based on origin IP
I want to be able to route UDP packets to different upstreams based on the packet's origin IP. With TCP this can be accomplished using `filter_chain_match`'s [`source_prefix_ranges`](https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/listener/v3/listener_components.proto#config-listener-v3-filterchainmatch), but as far as I can tell, filter chains cannot be applied to UDP listeners. It would be great to be able to have some sort of mechanism to achieve this with UDP, too.

### PR DESCRIPTION
**Pull request by zhxie**
Signed-off-by: Xie Zhihao zhihao.xie@intel.com

Commit Message: udp: add router for UDP proxy
Additional Description: match and route UDP datagrams to different clusters
Risk Level: Medium
Testing: Unit and integration
Docs Changes: UDP proxy example configuration
Release Notes: Added
Platform Specific Features: N/A
Fixes #17652

### CODE REVIEW

File: source/extensions/filters/udp/udp_proxy/router/router_impl.cc
Comment (zhxie) [reply source/extensions/filters/udp/udp_proxy/router/router_impl.cc:41]: Ok, I will remove these data inputs, but I still think the validation is necessary since a HTTP data input is also syntax valid for the UDP router and the validator will keep the gate.

File: generated_api_shadow/envoy/extensions/filters/udp/udp_proxy/v3/route.proto
Comment (zhxie) [reply generated_api_shadow/envoy/extensions/filters/udp/udp_proxy/v3/route.proto:1]: Ok.

File: source/extensions/filters/udp/udp_proxy/router/router.h
Comment (zhxie) [reply source/extensions/filters/udp/udp_proxy/router/router.h:24]: Changed, thanks.

**Comment by mattklein123** [comment]
Please check CI. Also is this still draft? I assume it's ready for review officially? Thanks.

/wait

**Comment by zhxie** [comment]
Yes, I will mark it ready for review after I fix all errors in CI.

File: source/extensions/filters/udp/udp_proxy/router/router_impl.cc
Comment (snowp) [reply source/extensions/filters/udp/udp_proxy/router/router_impl.cc:41]: The way the factory registrations work is that registrations are unique per data input type, so since you're using the NetworkDataInput it won't share registrations with HttpDataInput registrations

Comment (zhxie) [reply source/extensions/filters/udp/udp_proxy/router/router_impl.cc:41]: https://github.com/envoyproxy/envoy/blob/818419918a6b9e05f9fdb42302c84ed4c784607a/test/extensions/filters/udp/udp_proxy/router_impl_test.cc#L218
Yes, I also think matcher is expected to work that way, but the test shows Envoy will not stop until the validation visitor returns the error. I guess the registration of `HttpDataInput` is shared with `NetworkDataInput`, do I miss something?

File: api/envoy/extensions/filters/udp/udp_proxy/v3/udp_proxy.proto
Comment (zhxie) [inline api/envoy/extensions/filters/udp/udp_proxy/v3/udp_proxy.proto:61]: I would like to not deprecate `cluster` since generic matcher API is working in progress and the current syntax is too complicated for routing to a single cluster. What is your opinion @mattklein123 ?

Comment (mattklein123) [reply api/envoy/extensions/filters/udp/udp_proxy/v3/udp_proxy.proto]: I would like to hear from @snowp on this. I think there should be some way of doing an any match which should not be very verbose.

File: envoy/network/filter.h
Comment (mattklein123) [inline envoy/network/filter.h:13]: nit: use full path for include.

File: source/extensions/filters/udp/udp_proxy/router/router_impl.cc
Comment (mattklein123) [inline source/extensions/filters/udp/udp_proxy/router/router_impl.cc:80]: It might be easier to just return a non-reference here. If you want to return a reference, you can avoid a double set lookup below by doing the find() and if not end() just return it directly. Then again, it should never be end() so you could just assert.

File: api/envoy/extensions/filters/udp/udp_proxy/v3/udp_proxy.proto
Comment (mattklein123) [inline api/envoy/extensions/filters/udp/udp_proxy/v3/udp_proxy.proto:56]: I still think we should deprecate this, and I would like to close out the comment thread and how to have a trivial generic matching config that would just always route to a single cluster, and use that in all of the examples. @snowp can you help with this thread of this PR? Thank you.

**Review by mattklein123 — CHANGES_REQUESTED** [review]
Thanks looking good. A few high level comments before continuing.

/wait

File: api/envoy/extensions/filters/udp/udp_proxy/v3/udp_proxy.proto
Comment (zhxie) [reply api/envoy/extensions/filters/udp/udp_proxy/v3/udp_proxy.proto:56]: https://github.com/envoyproxy/envoy/blob/0458877b46f9bcb96f843331e2515f7aaf6bb800/api/envoy/config/common/matcher/v3/matcher.proto#L130-L144
I have a suggestion, shall we remove the requirement of this `matcher_type`, and change the name of `on_no_match` to `default_match`. Then we can have an only `default_match` for routing to a single cluster.

Comment (snowp) [reply api/envoy/extensions/filters/udp/udp_proxy/v3/udp_proxy.proto:56]: I think perhaps removing the requirement of a match would work yeah. We can't just rename fields, but we could potentially go through a deprecation cycle if we wanted (cc @envoyproxy/api-shepherds)

Also I just realized we want to be using the https://github.com/cncf/xds/blob/main/xds/type/matcher/v3/matcher.proto proto here, not the old config/common/matcher one

Comment (snowp) [reply api/envoy/extensions/filters/udp/udp_proxy/v3/udp_proxy.proto:56]: Alternatively we could extend SinglePredicate to have an always true branch:

```
      oneof match_type {
        option (validate.required) = true;

        // A single predicate to evaluate.
        SinglePredicate single_predicate = 1;

        // A list of predicates to be OR-ed together.
        PredicateList or_matcher = 2;

        // A list of predicates to be AND-ed together.
        PredicateList and_matcher = 3;

        // The invert of a predicate
        Predicate not_matcher = 4;
      }
```

e.g. a `bool match_all` oneof option. Perhaps a little less elegant but it would also do the trick

Comment (zhxie) [reply api/envoy/extensions/filters/udp/udp_proxy/v3/udp_proxy.proto:56]: So, I think we shall switch to xds/matcher, and use `matcher_list` instead of `matcher_tree` in the example to make it simpler. And we can remove the requirement in the future.

Comment (zhxie) [reply api/envoy/extensions/filters/udp/udp_proxy/v3/udp_proxy.proto:56]: Ok, I have tested with `matcher_list`, but I think it is much more verbose then `matcher_tree`

```yaml
matcher_list:
  matchers:
  - predicate:
      single_predicate:
        input:
          name: source-ip
          typed_config:
            '@type': type.googleapis.com/envoy.type.matcher.v3.SourceIpMatchInput
        value_match:
          exact: "127.0.0.1"
    on_match:
      action:
        name: route
        typed_config:
          '@type': type.googleapis.com/envoy.extensions.filters.udp.udp_proxy.v3.Route
          cluster: service_udp

matcher_tree:
  input:
    name: source-ip
    typed_config:
      '@type': type.googleapis.com/envoy.type.matcher.v3.SourceIpMatchInput
  exact_match_map:
    map:
      "127.0.0.1":
        action:
          name: route
          typed_config:
            '@type': type.googleapis.com/envoy.extensions.filters.udp.udp_proxy.v3.Route
            cluster: service_udp
```

File: envoy/network/filter.h
Comment (zhxie) [reply envoy/network/filter.h]: Ok.

File: source/extensions/filters/udp/udp_proxy/router/router_impl.cc
Comment (zhxie) [reply source/extensions/filters/udp/udp_proxy/router/router_impl.cc]: Ok, update to return a non-reference value.

File: api/envoy/extensions/filters/udp/udp_proxy/v3/udp_proxy.proto
Comment (zhxie) [reply api/envoy/extensions/filters/udp/udp_proxy/v3/udp_proxy.proto]: I have re-deprecated the original `cluster`. So shall we use the `matcher_tree` or wait the xDS API update for a concise syntax like remove the requirement of `matcher_tree` and `matcher_list`.

Comment (mattklein123) [reply api/envoy/extensions/filters/udp/udp_proxy/v3/udp_proxy.proto]: I think both of these are still verbose if we just want to universally always route to a cluster without any matching effectively. @snowp what do you recommend here? Can we just make the matcher change now to support something less verbose for a universal match?

/wait

Comment (snowp) [reply api/envoy/extensions/filters/udp/udp_proxy/v3/udp_proxy.proto]: Started a PR here https://github.com/cncf/xds/pull/19 ptal

**Comment by mattklein123** [comment]
What's the status of this PR? Are we still waiting for the simplified xDS matching option to land? cc @snowp 

/wait-any

**Comment by zhxie** [comment]
Yes, I am waiting for the PR cncf/xds#19 being merged for a simplified xDS matching option for a single cluster.

**Comment by zhxie** [comment]
Wait for #19143 

/wait

**Comment by zhxie** [comment]
The PR is not staled, and further work is needed after merging the required PR.

**Comment by jmarantz** [comment]
please fix format; thanks!

/wait

File: source/extensions/filters/udp/udp_proxy/router/router.h
Comment (RyanTheOptimist) [inline source/extensions/filters/udp/udp_proxy/router/router.h:17]: It would be nice to have a short high level explanation of what a Router does in this context. Presumably it looks as a source_address and returns a cluster name to send that packet to?

File: source/extensions/filters/udp/udp_proxy/router/router.h
Comment (RyanTheOptimist) [inline source/extensions/filters/udp/udp_proxy/router/router.h:23]: Is "address" a source or a destination address? (Presumably source. If so source_address would be a nice name for this parameter).

What's the return value? Presumably a cluster name? It'd be nice to add a comment to this effect.

File: source/extensions/filters/udp/udp_proxy/router/router_impl.cc
Comment (RyanTheOptimist) [inline source/extensions/filters/udp/udp_proxy/router/router_impl.cc:71]: nit: To remove a level of nesting in the code, since the body of the "if" returns, we can remove the "else" and out-dent everything below.

File: source/extensions/filters/udp/udp_proxy/udp_proxy_filter.h
Comment (RyanTheOptimist) [inline source/extensions/filters/udp/udp_proxy/udp_proxy_filter.h:94]: Should this perhaps be named "clusterForAddress()" or maybe "route()"?

**Review by RyanTheOptimist — COMMENTED** [review]
Do you happen to know why one of the tests is failing?

File: source/common/network/matching/data_impl.h
Comment (zhxie) [reply source/common/network/matching/data_impl.h:14]: In fact, this method refers to the [HttpMatchingDataImpl](https://github.com/envoyproxy/envoy/blob/main/source/common/http/matching/data_impl.h#L16).

File: envoy/network/filter.h
Comment (zhxie) [reply envoy/network/filter.h:216]: Ok

File: source/extensions/filters/udp/udp_proxy/router/router.h
Comment (zhxie) [reply source/extensions/filters/udp/udp_proxy/router/router.h]: Ok.

File: source/extensions/filters/udp/udp_proxy/router/router.h
Comment (zhxie) [reply source/extensions/filters/udp/udp_proxy/router/router.h:28]: Ok.

File: source/extensions/filters/udp/udp_proxy/router/router_impl.cc
Comment (zhxie) [reply source/extensions/filters/udp/udp_proxy/router/router_impl.cc]: Ok.

File: source/extensions/filters/udp/udp_proxy/udp_proxy_filter.h
Comment (zhxie) [reply source/extensions/filters/udp/udp_proxy/udp_proxy_filter.h]: Ok, renamed to `route`.

File: source/common/network/matching/data_impl.h
Comment (RyanTheOptimist) [reply source/common/network/matching/data_impl.h:14]: How is this method used, though? Where is it called?

File: source/extensions/filters/udp/udp_proxy/router/router.h
Comment (RyanTheOptimist) [reply source/extensions/filters/udp/udp_proxy/router/router.h:28]: Let's rename "address" to "source address" in the parameter. (Here and elsewhere)

**Review by RyanTheOptimist — COMMENTED** [review]
Looks like you'll need to do a main merge to fix the release notes conflict since the release has happened.

This basically look good to me (modulo two small questions). But I think snow should definitely take a look since they had thoughts on an earlier version of this PR.

File: source/extensions/filters/udp/udp_proxy/router/router.h
Comment (zhxie) [reply source/extensions/filters/udp/udp_proxy/router/router.h:28]: I will change these names.

File: source/common/network/matching/data_impl.h
Comment (zhxie) [reply source/common/network/matching/data_impl.h:14]: I am not sure how is the `name()` in these `MatchingData` used. Do you have any idea? @snowp

File: api/envoy/extensions/filters/udp/udp_proxy/v3/udp_proxy.proto
Comment (zhxie) [reply api/envoy/extensions/filters/udp/udp_proxy/v3/udp_proxy.proto]: Since related PRs have been merged, we can now write
```
on_no_match:
  action:
    name: route
    typed_config:
      '@type': type.googleapis.com/envoy.extensions.filters.udp.udp_proxy.v3.Route
      cluster: service_udp
```
to serve a single UDP cluster. And it is much simpler than the previous one.

**Comment by zhxie** [comment]
Since we have merged #19493 for network inputs, the UDP router would be better to comply with it. I meet with some problems when I fixed the merging conflicts and wish to get your thoughts.
#19493 introduced network inputs, but it is majorly for stream network since it requires a `socket` to generate inputs like source IP and destination IP. But in UDP, we do not own any socket.
So, shall we create a new matching data like `UdpMatchingData` (maybe easier), or update the current existing `MatchingData` for TCP and UDP compatible? @snowp @mattklein123

**Comment by zhxie** [comment]
I have submitted a patch for the former option. See #20116

File: source/extensions/filters/udp/udp_proxy/udp_proxy_filter.h
Comment (RyanTheOptimist) [inline source/extensions/filters/udp/udp_proxy/udp_proxy_filter.h:117]: It looks like the way this member is used, it can be an object instead of a shared pointer. Would that work?

File: source/extensions/filters/udp/udp_proxy/udp_proxy_filter.h
Comment (RyanTheOptimist) [inline source/extensions/filters/udp/udp_proxy/udp_proxy_filter.h:353]: nit: Please comment on what the key in this map is. (Cluster name?)

File: source/extensions/filters/udp/udp_proxy/udp_proxy_filter.cc
Comment (RyanTheOptimist) [inline source/extensions/filters/udp/udp_proxy/udp_proxy_filter.cc:46]: I believe the type of route is std::string. Please make that explicit instead of using auto.

File: source/extensions/filters/udp/udp_proxy/udp_proxy_filter.h
Comment (zhxie) [reply source/extensions/filters/udp/udp_proxy/udp_proxy_filter.h:355]: Ok, done.

File: source/extensions/filters/udp/udp_proxy/udp_proxy_filter.cc
Comment (zhxie) [reply source/extensions/filters/udp/udp_proxy/udp_proxy_filter.cc]: Ok, done.

File: source/extensions/filters/udp/udp_proxy/udp_proxy_filter.h
Comment (zhxie) [reply source/extensions/filters/udp/udp_proxy/udp_proxy_filter.h:118]: In fact, I do not think so. `Router` is abstract, so we cannot use its object here.

File: api/envoy/extensions/filters/udp/udp_proxy/v3/udp_proxy.proto
Comment (mattklein123) [inline api/envoy/extensions/filters/udp/udp_proxy/v3/udp_proxy.proto:64]: Can you ref link back to the docs to where this is described and there are examples? People are not going to know how to fill this out.

File: api/envoy/extensions/filters/udp/udp_proxy/v3/udp_proxy.proto
Comment (mattklein123) [inline api/envoy/extensions/filters/udp/udp_proxy/v3/udp_proxy.proto:57]: Please update the docs here to mention this is deprecated and to use `matcher` instead.

**Review by mattklein123 — CHANGES_REQUESTED** [review]
Thanks this LGTM at a high level. Can you merge main and then I can do a full pass? Thanks.

/wait

File: api/envoy/extensions/filters/udp/udp_proxy/v3/udp_proxy.proto
Comment (zhxie) [reply api/envoy/extensions/filters/udp/udp_proxy/v3/udp_proxy.proto:67]: Done.

File: api/envoy/extensions/filters/udp/udp_proxy/v3/udp_proxy.proto
Comment (zhxie) [reply api/envoy/extensions/filters/udp/udp_proxy/v3/udp_proxy.proto:58]: Ok, done.

File: docs/root/version_history/current.rst
Comment (mattklein123) [inline docs/root/version_history/current.rst:145]: Please add a note in deprecated below for the deprecated field.

File: source/extensions/filters/udp/udp_proxy/router/router.h
Comment (mattklein123) [inline source/extensions/filters/udp/udp_proxy/router/router.h:33]:
```suggestion
   * Returns all cluster names in the router. The UDP proxy filter requires every cluster name for
```

File: source/extensions/filters/udp/udp_proxy/router/router_impl.cc
Comment (mattklein123) [inline source/extensions/filters/udp/udp_proxy/router/router_impl.cc:39]: I think it would be a bit cleaner to just synthesize matcher config for this case. Then you can use unified code for the rest of it? (Basically it makes the deprecated path cheaper to maintain and understand as the only special case is here.)

File: source/extensions/filters/udp/udp_proxy/router/router_impl.cc
Comment (mattklein123) [inline source/extensions/filters/udp/udp_proxy/router/router_impl.cc:63]: Technically isn't it possible to send UDP/datagrams over a pipe? Or do we block this in config? Either way it seems like this should be handled elsewhere as a config load issue vs. on every route request?

File: source/extensions/filters/udp/udp_proxy/udp_proxy_filter.cc
Comment (mattklein123) [inline source/extensions/filters/udp/udp_proxy/udp_proxy_filter.cc:41]: nit: this log is now misleading. I would see if the cluster is used before logging and erasing? If you store an iterator the perf delta will be negligible.

File: source/extensions/filters/udp/udp_proxy/router/router.h
Comment (mattklein123) [inline source/extensions/filters/udp/udp_proxy/router/router.h:30]: nit: I would pass by const ref to avoid shared_ptr inc/dec

**Review by mattklein123 — CHANGES_REQUESTED** [review]
Thanks very nice. Flushing out some small comments.

/wait

File: docs/root/version_history/current.rst
Comment (zhxie) [reply docs/root/version_history/current.rst:145]: Ok, added.

File: source/extensions/filters/udp/udp_proxy/router/router.h
Comment (zhxie) [reply source/extensions/filters/udp/udp_proxy/router/router.h:33]: Thanks, changed.

File: source/extensions/filters/udp/udp_proxy/router/router_impl.cc
Comment (zhxie) [reply source/extensions/filters/udp/udp_proxy/router/router_impl.cc:63]: Yes, it is redundant. I will remove it, thanks.

File: source/extensions/filters/udp/udp_proxy/udp_proxy_filter.cc
Comment (zhxie) [reply source/extensions/filters/udp/udp_proxy/udp_proxy_filter.cc:41]: Thanks, fixed.

File: source/extensions/filters/udp/udp_proxy/router/router.h
Comment (zhxie) [reply source/extensions/filters/udp/udp_proxy/router/router.h:30]: Ok, I will make them `const`.
By the way, `InstanceConstSharedPtr` is a `shared_ptr<const Instance>`. I think it is not necessary to change to a const ref since the `const Instance` cannot be changed.

File: source/extensions/filters/udp/udp_proxy/router/router_impl.cc
Comment (zhxie) [reply source/extensions/filters/udp/udp_proxy/router/router_impl.cc:39]: I am sorry but I cannot catch your idea here. Do you mention we can move `cluster_` out of `Router`?

Comment (zhxie) [reply source/extensions/filters/udp/udp_proxy/router/router_impl.cc:39]: Ok, I catch your thought finally, I will submit the commit soon.

File: source/extensions/filters/udp/udp_proxy/router/router.h
Comment (mattklein123) [inline source/extensions/filters/udp/udp_proxy/router/router.h:29]: Sorry I meant `const Network::Address::Instance&`

**Review by mattklein123 — CHANGES_REQUESTED** [review]
Thanks LGTM other than small comment and fixing CI.

/wait

File: source/extensions/filters/udp/udp_proxy/router/router.h
Comment (zhxie) [reply source/extensions/filters/udp/udp_proxy/router/router.h:29]: https://github.com/envoyproxy/envoy/blob/eede13d30efceb3e8c9fb3bb3e39a527495a86f9/source/extensions/filters/udp/udp_proxy/udp_proxy_filter.cc#L50

In the caller method, we will get a `InstanceConstSharedPtr`. If we change the `Router::route`'s signature to `const Network::Address::Instance&`, we have to change the code in the caller to

```cpp
const std::string& route = config_->route(*data.addresses_.local_, *data.addresses_.peer_); 
```

Do you think it is acceptable?

Comment (zhxie) [reply source/extensions/filters/udp/udp_proxy/router/router.h]: Fixed, I have updated all methods in the caller chain.

**Comment by zhxie** [comment]
It is strange that the patch does not touch any codes in `source/common/http` but the CI complains low coverage on it.

/retest

**Comment by zhxie** [comment]
Flaky macOS test.

/retest

**Review by mattklein123 — APPROVED** [review]
Thank you. Great work!

**Pull request merged**
